### PR TITLE
!feature: refactor mesh for GL330

### PR DESCRIPTION
**Pull request by pollend**
I've generalized mesh data a bit with this layout configuration. you specific the type of layout and the index that is used by the shader. This also forces us to use the opengl 3.3 version of GLSL.  This does away with the fixed function pipeline for meshes so most of these shaders will need to be rewritten so that they correctly take advantage of the refactored mesh. moving away from the fixed function pipeline should make it more feasible to take advantage of vulkan if we want to adopt that at some point.  

I think I would like to define some standards for shaders: 

### StandardMeshData

```
layout (location = 0) in vec3 in_vert;
layout (location = 1) in vec3 in_normal;
layout (location = 2) in vec2 in_uv0;
layout (location = 3) in vec2 in_uv0;
layout (location = 4) in vec4 in_color0;
layout (location = 5) in vec3 in_light0;
```

### predefined uniforms

- `tera_model_view`
- `tera_normal`
- `tera_projection`

### CODE REVIEW

File: engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java [hunk 1]
```diff
@@ -93,188 +65,111 @@ public TFloatList getVertices() {
...
-        for (int v = 0; v < vertexCount; ++v) {
-            for (int partIndex = 0; partIndex < parts.size(); ++partIndex) {
-                TFloatIterator part = parts.get(partIndex);
-                for (int i = 0; i < partSizes.get(partIndex); ++i) {
-                    vertexBuffer.put(part.next());
-                }
+        ByteBuffer buffer = BufferUtils.createByteBuffer(stride * newData.getSize());
+        for (int x = 0; x < newData.getSize(); x++) {
+            for (Layout layout : targets) {
+                layout.write(x, buffer);
```
Comment (4Denthusiast) [inline engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java:102]: Shouldn't this be taking the stride into account too?

Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java:102]: I pack all the data local to each other

XYZRGB|XYZRGB|XYZRGB|XYZRGB

stride is the length of the payload so XYZRGB. the offset into the same position for a position in the chunk is the lenght of the payload.

I could of packed it like this.

XYZXYZXYZXYZ | RGBRGBRGBRGB. the stride would be zero and the offset would be some number into the assoicated sections.

File: engine/src/main/java/org/terasology/engine/rendering/nui/internal/LwjglCanvasRenderer.java [hunk 2]
```diff
@@ -100,6 +88,9 @@
     private DisplayDevice displayDevice;
     private float uiScale = 1f;
 
+    private Matrix4fStack modelMatrixStack = new Matrix4fStack(100000);
```
Comment (pollend) [inline engine/src/main/java/org/terasology/engine/rendering/nui/internal/LwjglCanvasRenderer.java:91]: umm used an arbitrarily large number for the max stack size.

File: engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java (hunk 1 above)
Comment (4Denthusiast) [reply engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java:102]: Oh, okay. I missed the bit about it just adding the data to the end of the buffer.

File: engine/src/main/java/org/terasology/engine/rendering/nui/internal/LwjglCanvasRenderer.java (hunk 2 above)
Comment (4Denthusiast) [reply engine/src/main/java/org/terasology/engine/rendering/nui/internal/LwjglCanvasRenderer.java:91]: This will actually allocate 100000 matrices. That's a few MB of space (and also, having lots of small objects may worsen garbage-collection performance even if they are all long-lived, although I don't know whether that still true if each of them is pretty self-contained like this).

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/ingame/metrics/DebugOverlay.java [hunk 3]
```diff
@@ -236,6 +237,11 @@ protected boolean isEscapeToCloseAllowed() {
         return false;
     }
 
+    @Override
+    public void onDraw(Canvas canvas) {
```
Comment (4Denthusiast) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/ingame/metrics/DebugOverlay.java:241]: Why is this necessary? It looks like it should just be equivalent to inheriting the implementation.

**Review by 4Denthusiast — COMMENTED** [review]
I've only looked at the code, not tested it. It mostly looks fine, but there were some minor issues.

File: engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java [hunk 4]
```diff
@@ -93,188 +71,126 @@ public TFloatList getVertices() {
...
+
+        // bind vertex array and buffer
+        GL30.glBindVertexArray(this.disposalAction.vao);
+        GL30.glBindBuffer(GL30.GL_ARRAY_BUFFER, this.disposalAction.vbo);
+        GL30.glBufferData(GL30.GL_ARRAY_BUFFER, buffer, GL30.GL_STATIC_DRAW);
+
+        int offset = 0;
+        for (Layout layout : targets) {
+            GL30.glEnableVertexAttribArray(layout.location);
+            if ((layout.flag & Layout.FLOATING_POINT) > 0) {
```
[the remark this answers is not in the document]
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java:121]: I have some plans to do a second pass on this. I want to split the layout into a nested class within a definition. this is actually better for memory performance: https://anteru.net/blog/2016/storing-vertex-data-to-interleave-or-not-to-interleave/ I was also looking a bit UnrealEngine to see how it manages buffers and they keep things non interleaved. 

Vertex definition would contain the memory and the attribute will define how the data is mapped into a buffer. we can have version flag so only the vertex data that is changed will be mapped into the VBO. I want to rework this a again with some research I've been doing on the side as I was setting this up. we have to handle the case where the size of the definition changes so all the buffered data will have to be re-written. still debating if it would be better to have one VBO or multiple VBO's. GL30 has a subdata function where you can save into part of the buffer and not attempt to overwrite the whole thing.  One thing I want to do in particular is have the data written straight to the buffer and I think we could have a flag that would let us save some of the data to some kind of backing array if that is important. 

- VertexDefintion (Position Color)
  - VertexAttribute (Position)
  - VertexAttribute (Color)
- VertexDefintion (UV0) 
  - VertexAttribute () 
- VertexDefintion (UV1)
  - VertexAttribute

File: engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java [hunk 5]
```diff
@@ -93,188 +71,126 @@ public TFloatList getVertices() {
...
-        if (disposalAction.vboIndexBuffer == 0) {
-            disposalAction.vboIndexBuffer = disposalAction.bufferPool.get(getUrn().toString());
-        }
-        VertexBufferObjectUtil.bufferVboElementData(disposalAction.vboIndexBuffer, indexBuffer, GL15.GL_STATIC_DRAW);
-        indexBuffer.flip();
+        GL30.glBindBuffer(GL30.GL_ELEMENT_ARRAY_BUFFER, this.disposalAction.ebo);
+        GL30.glBufferData(GL30.GL_ELEMENT_ARRAY_BUFFER, bufferIndices, GL30.GL_STATIC_DRAW);
+        indexCount = indices.size();
+
+        GL30.glBindVertexArray(0);
```
[the remark this answers is not in the document]
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java:158]: I think you would just end up unbinding the VBO from the VAO.

File: engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import gnu.trove.list.TFloatList;
+import gnu.trove.list.array.TFloatArrayList;
+import org.terasology.engine.rendering.assets.mesh.layout.FloatLayout;
+import org.terasology.engine.rendering.assets.mesh.layout.Layout;
+
+public class StandardMeshData extends MeshData {
+    public static final int VERTEX_INDEX = 0;
+    public static final int NORMAL_INDEX = 1;
+    public static final int UV0_INDEX = 2;
+    public static final int UV1_INDEX = 3;
```
[the remark this answers is not in the document]
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java:15]: I have a second UV texture unit reserved. I think the old code also have a second UV1 but it was used for light data :confused:.

[the remark this answers is not in the document]
Comment (4Denthusiast) [reply engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java:15]: In the case of normals at least, that would normally use the same coordinates, just with a different texture.

[the remark this answers is not in the document]
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java:15]: you could also use the same texture with multiple UV's but I see your point. I was trying to keep consistent with the old code.

[the remark this answers is not in the document]
Comment (4Denthusiast) [reply engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java:15]: The old code was only using UV1 for something hacky that's no longer needed though. Removing it entirely seems like a natural part of this change.

[the remark this answers is not in the document]
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java:15]: sounds good to me.

File: engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java [hunk 7]
```diff
@@ -12,10 +12,8 @@
...
     public static final int UV0_INDEX = 2;
-    public static final int UV1_INDEX = 3;
-    public static final int COLOR0_INDEX = 4;
-    public static final int LIGHT0_INDEX = 5;
-
+    public static final int COLOR0_INDEX = 3;
+    public static final int LIGHT0_INDEX = 4;
 
     public final TFloatList vertices;
     public final TFloatList uv0;
```
Comment (4Denthusiast) [inline engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java:19]: You missed the actual `uv1` field.

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/ingame/metrics/DebugOverlay.java (hunk 3 above)
Comment (4Denthusiast) [reply engine/src/main/java/org/terasology/engine/rendering/nui/layers/ingame/metrics/DebugOverlay.java]: The `Canvas` import isn't needed either.

File: engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/assets/mesh/StandardMeshData.java]: actually UV1 is used for GLTF.

File: engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java [hunk 8]
```diff
@@ -93,188 +71,126 @@ public TFloatList getVertices() {
...
-        }
-    }
-
     @Override
     public void render() {
         if (!isDisposed()) {
-            preRender();
-            doRender();
-            postRender();
+            GL30.glBindVertexArray(disposalAction.vao);
```
[the remark this answers is not in the document]
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java:77]: how does the disposable action work? I was just following what was already written. would it be disposed by the buffer and not the mesh itself.

File: engine/src/main/java/org/terasology/engine/rendering/nui/internal/LwjglCanvasRenderer.java [hunk 9]
```diff
@@ -100,6 +86,9 @@
     private DisplayDevice displayDevice;
     private float uiScale = 1f;
 
+    private Matrix4fStack modelMatrixStack = new Matrix4fStack(10000);
```
[the remark this answers is not in the document]
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/nui/internal/LwjglCanvasRenderer.java:89]: what would be a good size. I guess we probably won't see a nui depth of over 10,000 entries. close to maybe 1000 so would 5000 be fair?

[the remark this answers is not in the document]
Comment (4Denthusiast) [reply engine/src/main/java/org/terasology/engine/rendering/nui/internal/LwjglCanvasRenderer.java:89]: I'd go with 1000 I think. I'm not all that familiar with the use case but I wouldn't expect it to ever have a stack of transformations that deep.

[the remark this answers is not in the document]
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/rendering/nui/internal/LwjglCanvasRenderer.java]: done, it looks good :+1:

File: engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (4Denthusiast) [reply engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java:122]: I tried to test this but it turned out to be a huge pain to set up new shaders, a new material, a new sort of mesh and a prefab to use that mesh, just to get one integer attribute, so I gave up. The code looks correct, so I'll just accept it without testing this bit.

**Review by 4Denthusiast — APPROVED** [review]
I ended up not testing the other variants of attribute binding (with the data either on the Java side or on the GLSL side being integers), but I tested that it's working with the standard meshes at least.
